blockchain/upgrade.go

```diff
@@ -1789,6 +1789,1162 @@ func upgradeToVersion8(ctx context.Context, db database.DB, dbInfo *databaseInfo
+			// follow.  When both output 0 and 1 are spent, it encodes N-1.
+			output0Unspent := code&0x01 != 0
+			output1Unspent := code&0x02 != 0
+			numBitmapBytes := code >> 2
+			if !output0Unspent && !output1Unspent {
+				numBitmapBytes++
+			}
+
+			// Ensure there are enough bytes left to deserialize the unspentness
+			// bitmap.
+			if uint64(len(oldSerialized[offset:])) < numBitmapBytes {
+				return errDeserialize("unexpected end of data for " +
+					"unspentness bitmap")
+			}
+
+			// Add sparse outputs for unspent outputs 0 and 1 as needed based on the
+			// details provided by the header code.
+			var outputIndexes []uint32
+			if output0Unspent {
+				outputIndexes = append(outputIndexes, 0)
+			}
+			if output1Unspent {
+				outputIndexes = append(outputIndexes, 1)
+			}
+
+			// Decode the unspentness bitmap adding a sparse output for each unspent
+			// output.
+			for i := uint32(0); i < uint32(numBitmapBytes); i++ {
+				unspentBits := oldSerialized[offset]
+				for j := uint32(0); j < 8; j++ {
+					if unspentBits&0x01 != 0 {
+						// The first 2 outputs are encoded via the
+						// header code, so adjust the output number
+						// accordingly.
+						outputNum := 2 + i*8 + j
+						outputIndexes = append(outputIndexes, outputNum)
+					}
+					unspentBits >>= 1
+				}
+				offset++
+			}
+
+			// Create a map to hold all of the converted outputs for the entry.
+			type convertedOut struct {
+				compressedAmount uint64
+				compressedScript []byte
+				scriptVersion    uint64
+			}
+			outputs := make(map[uint32]*convertedOut)
+
+			// Decode and add all of the outputs.
+			for _, outputIndex := range outputIndexes {
+				// Deserialize the compressed amount and ensure there are bytes
+				// remaining for the compressed script.
+				compressedAmount, bytesRead := deserializeVLQ(oldSerialized[offset:])
+				if bytesRead == 0 {
+					return errDeserialize("unexpected end of data during decoding " +
+						"(compressed amount)")
+				}
+				offset += bytesRead
+
+				// Decode the script version.
+				scriptVersion, bytesRead := deserializeVLQ(oldSerialized[offset:])
+				if bytesRead == 0 {
+					return errDeserialize("unexpected end of data during decoding " +
+						"(script version)")
+				}
+				offset += bytesRead
+
+				// Decode the compressed script size and ensure there are enough bytes
+				// left in the slice for it.
+				const compressionVersion = 1
+				scriptSize := decodeCompressedScriptSize(oldSerialized[offset:],
+					compressionVersion)
+				// Note: scriptSize == 0 is OK (an empty compressed script is valid)
+				if scriptSize < 0 {
+					return errDeserialize("negative script size")
+				}
+				if len(oldSerialized[offset:]) < scriptSize {
+					return errDeserialize(fmt.Sprintf("unexpected end of "+
+						"data after script size (got %v, need %v)",
+						len(oldSerialized[offset:]), scriptSize))
+				}
+				compressedScript := make([]byte, scriptSize)
```

Comment:
It probably won't make much difference in the overall time and/or memory given most entries are pretty small, but the script really doesn't need to be duplicated here.  It's only duplicated in the normal path because the utxo is later returned and thus no reference into the data can be kept so the serialized data can be GCd, but that doesn't apply here because it's only live for the duration of a single record.

```suggestion
				compressedScript := oldSerialized[offset:offset+scriptSize]
```

Reply:
Good call, updated.